### DISCUSSION TITLE
PKCS11 and PSK

### DISCUSSION THREAD

**Discussion by satishkrishna687** [strongswan/strongswan#1592] [Q&A] [answered]
I'm trying to establish a Host-to-Host connection using Pre-Shared Key (PSK) obtained from a Hardware Security Module (SoftHSM in my case) or OP-TEE device. I have been sucessful in making it work RSA key pair with certificates like in **case1** in the below image. I've attempted to import an AES key and a text file into the softHSM, and provided the corresponding handle number in the 'secrets' section of my swanctl.conf file. However, I keep encountering an error message that reads 'failed to locate private key at <handle_no>' show in **case 3**. Is there any workaround to make strongswan to authenticate with psk reterived from softhsm like in **case 3**. I haven't checked it out with OP-TEE, can I make my above requirement work if I change to OP-TEE like in **case 2 and Case 4**. Is there any work around for me to make strongswan work on PSK from softhsm or OP-TEE.

<a href="https://freeimage.host/i/HX5H4j9"><img src="https://iili.io/HX5H4j9.md.png" alt="HX5H4j9.md.png" border="0"></a><br /><a target='_blank' href='https://freeimage.host/'>free image hosting</a><br />

**Comment by satishkrishna687** [comment]
@tobiasbrunner please help me out. I have to get this thing working somehow.

**Comment by tobiasbrunner** [comment, marked as answer]
strongSwan does currently not support retrieving PSKs from tokens via PKCS#11. It can only access RSA/ECDSA keys.

**Reply by satishkrishna687** [reply]
@tobiasbrunner Thank you so much. 
I know this is kinda repetative but is there anyway I can use Openssl to access psk form my OP-TEE in **OPTION 1**  or
**OPTION 2** is it possile if I use Openssl's Engine API[PKCS#11] to access PSK from my OP-TEE to establish host to host connection. Please let me know if there is a workaround to this problem via openssl or through any other means.
<a href="https://freeimage.host/i/HXjrw7a"><img src="https://iili.io/HXjrw7a.md.png" alt="HXjrw7a.md.png" border="0"></a>

**Reply by tobiasbrunner** [reply]
Not without code changes either as the _openssl_ plugin does not implement an API to provide PSKs.
